sw/device/lib/dif/dif_hmac.c

```diff
@@ -227,48 +200,121 @@ dif_hmac_result_t dif_hmac_irq_force(const dif_hmac_t *hmac,
   return kDifHmacOk;
 }
 
-dif_hmac_result_t dif_hmac_mode_hmac_start(const dif_hmac_t *hmac,
-                                           const uint8_t *key) {
+/**
+ * Sets up the CFG value for a given per-transaction configuration.
+ *
+ * This only sets the right values for the ENDIAN_SWAP / DIGEST_SWAP values,
+ * using the values in `config`.
+ *
+ * The implementation here is careful to only update `*device_config` once it
+ * has calculated the entire value for the register, rather than gradually
+ * updating it early. The value of `*device_config` is only updated if the
+ * function returns #kDifHmacOk.
+ *
+ * @param[inout] device_config HMAC CFG register value to be updated;
+ * @param config A per-transaction configuration.
+ * @returns #kDifHmacError if the config is invalid, #kDifHmacOk if
+ * `*device_config` was sucessfully updated.
+ */
+static dif_hmac_result_t dif_hmac_calculate_device_config_value(
+    uint32_t *device_config, const dif_hmac_transaction_config_t config) {
+  uint32_t new_config = *device_config;
+
+  bool swap_message_endianness;
+  bool swap_digest_endianness;
+
+  // Set the byte-order of the input message.
+  switch (config.message_endianness) {
+    case kDifHmacEndiannessBig:
+      swap_message_endianness = false;
+      break;
+    case kDifHmacEndiannessLittle:
+      swap_message_endianness = true;
+      break;
+    default:
+      return kDifHmacError;
+  }
+
+  // Set the byte-order of the digest.
+  switch (config.digest_endianness) {
+    case kDifHmacEndiannessBig:
+      swap_digest_endianness = false;
+      break;
+    case kDifHmacEndiannessLittle:
+      swap_digest_endianness = true;
+      break;
+    default:
+      return kDifHmacError;
+  }
+
+  new_config = bitfield_bit32_write(new_config, HMAC_CFG_ENDIAN_SWAP,
+                                    swap_message_endianness);
+  new_config = bitfield_bit32_write(new_config, HMAC_CFG_DIGEST_SWAP,
+                                    swap_digest_endianness);
```

Comment:
Could these be moved to just after the `switch` where the bool is initialized?

Reply:
No, this is explicitly covered in the description of the function: The value of `*device_config` won't be updated if this function returns `kDifHmacError`.